### ISSUE TITLE
Suggestion: JSONize bionic and mutation functions for further modding convenience.

### ISSUE THREAD

**Issue by jason0320** [CleverRaven/Cataclysm-DDA#13783]
I knew this requires work and not everyone has time for coding, but hard coded stuff are irritating to mod. Could someone kindly extract the functions of bionics and mutations for modders please?

**Comment by chaosvolt** [comment]
This is one I've been hoping for a while, but it would be a pain in the ass to implement. Same goes with artifact properties ( https://github.com/CleverRaven/Cataclysm-DDA/issues/13318 ), these could be useful for modding but aren't currently essential as-is.

Being able to add more bionic weapons would be delicious though. XP

**Comment by Coolthulhu** [comment]
Extract what?
Which functions specifically?

Stuff like bio-weapons (both from mutations and bionics) could become modable. Stat mods are already modable in mutations.
Stuff like the Magnet CBM will not, because it inherently depends on hardcoded functions.

**Comment by chaosvolt** [comment]
The "this bionic hands you a weapon when activated" would be the function that comes to mind first, yes. Armor values might work too, but I can't see as many immediate uses for them.

Mutations however have a lot more useful properties. We already have stat mods and body part wetness protection from them. Body part warmth modifiers and body part "can't wear clothes here" effects might also be JSONizable, I'm hoping.

**Comment by jason0320** [comment]
Sorry for not being specific. I mean if we can add complete new bionics and whole new functions like "charge UPS with nutrition" or just simply add new bio weapon like a bio nx-17.
BTW is nx-17's special charge function moddable?

**Comment by chaosvolt** [comment]
Hmm. I'm not sure how easy it'd be to add in such specific effects, a lot of the more clever stuff is still impractical to JSONize. Not even sure how a "lower stat X (such as nutrition) to fill item Y with its relevant ammo of type Z" effect would work code-wise, let alone how many uses that would have.

**Comment by Coolthulhu** [comment]
> I mean if we can add complete new bionics and whole new functions like "charge UPS with nutrition"

Nope. That depends on C++ code. Lua hooks for this stuff could exist, but it won't ever be doable with jsons alone.

> is nx-17's special charge function moddable?

Hardcoded. Making different charge rifles is possible, but they will still share a lot of functions with nx-17. The effect the rifle's charge has on its ammo is hardcoded, but rifle's stats can vary.

**Comment by jason0320** [comment]
OK thanks. Does that means the lost an arm effect of fusion blaster arm CBM is also hardcoded? So blaster can't be Iron Man but Megaman style? Loss an arm for a rifle does't sound good.

**Comment by chaosvolt** [comment]
Hmm. I would assume that loss of an arm COULD be shunted away into a simple flag, but that would be, well...who's going to have a use for that added functionality, outside of an Evil Dead mod? >.>

**Comment by Coolthulhu** [comment]
> Does that means the lost an arm effect of fusion blaster arm CBM is also hardcoded?

Yes, but this one wouldn't be that hard to de-hardcode. Also, laser finger CBM doesn't cost an arm.

**Comment by jason0320** [comment]
>Yes, but this one wouldn't be that hard to de-hardcode. Also, laser finger CBM doesn't cost an arm.

How? I'd be glad to know about that. :) And are tags like "LIGHTNING" and "BOUNCE" ranged only?

**Comment by Coolthulhu** [comment]
> How?

All de-hardcoding requires changes to hardcoded stuff. In this case probably adding flags to bionics and changing the arm removal thing into a flag rather than hardcoded id check.

> And are tags like "LIGHTNING" and "BOUNCE" ranged only?

Yes, but the range could be 1.

**Comment by chaosvolt** [comment]
Also, about the arm removal thing. Changing it to a bionic property with the option of specifying what body part(s) get removed would be more versatile.

Adding replacement limbs for a little real-world flavor would be one possible use, but the effect of "this bionic mimics some of the functions of the limb being replaces" might have to be hardcoded to some extent, unless "mimics function of this limb when installed" is also made a modular property.

**Comment by Midaychi** [comment]
Iunno. Bionics and stuff like that seem like they'd work better in lua than json. 
With json, you can only hardcode stuff in source and then refer to it in the json, while with a LUA hook all you have to do is make the game ask the lua interpreter 'wtf do' and 'here's a thing that happened', but lua still requires programming knowledge I guess.

**Comment by Rivet-the-Zombie** [comment]
We don't need to start requiring LUA knowledge to work on the primary code; folks already have to learn enough languages to contribute as it is.

**Comment by chaosvolt** [comment]
Agreed. JSONize as much as is practical to do so, and leave the rest in its existing code format. LUAifiying it instead would remove the benefit of simplifying things for mods and future mainline additions.

**Comment by i2amroy** [comment]
Indeed, while there are definitely going to be things that will never be JSON'd, there are plenty of things we can work on moving over. Anything dealing with needs, stuff that uses the effects system, there's still quite a bit that can be moved over to the JSONs.

**Comment by drbig** [comment]
I believe the idea is that once things become clear (they do so naturally) some basic functionality can be generalized enough to be parametrized in JSON (what a sentence here). It's one of these "yes we know; not now really" things.

**Comment by chaosvolt** [comment]
Basically something that will be useful for future features once exisisting priorities have been dealt with.

**Comment by Nandorianen** [comment]
The least (and easiest) that could be done is value exporting. Tweaking is good. Why is wet penalties tweakable, but cargo capacity is not? Getting the list of Integrated Toolset properties from json also sounds good. Armor, encumbrance, unusable clothing, range, damage, tool qualities and any other values or lists would be great to see in json.
